### DISCUSSION TITLE
Use go2rtc with home assistant remotely through cloudflare.

### DISCUSSION THREAD

**Discussion by Daniel-dev22** [blakeblackshear/frigate#4624] [Ask A Question]
I'm running the latest dev image and webrtc works locally however when I attempt to use it remotely over a cloudflare tunnel it doesn't work both with and without port forwarding.

I attempted to add a subdomain on the domain the companion app uses to connect and I pointed it to port 8555 on the machine running frigate/go2rtc/home assistant and that didn't work.

It works when I use a warp tunnel without any port forwarding.

Webrtc also works if I do a direct cloudflare tunnel to frigate:5000 but I'm trying to use webrtc through the home assistant companion app without a warp tunnel on.

Here's my tunnel configurations through the zero Trust Dashboard.

![Screenshot_20221204_105956_Chrome](https://user-images.githubusercontent.com/47092714/206021424-7ff14363-520c-478d-9f40-ffb6a6b9519b.jpg)
![Screenshot_20221206_160452_Chrome](https://user-images.githubusercontent.com/47092714/206023726-9b28620e-a8e0-4a34-a810-7071b786f7c5.jpg)
![Screenshot_20221206_161004_Chrome](https://user-images.githubusercontent.com/47092714/206023729-52df3697-8888-47e6-a60c-330654b41fa6.jpg)

My OPNsense port forwarding attempt. I also checked to make sure the rule appeared in WAN and it did.

The cloudflare_ip_addresses is an alias to this 
https://www.cloudflare.com/ips-v4

![Screenshot_20221206_142613_Chrome](https://user-images.githubusercontent.com/47092714/206021516-f6d77fa7-baee-4804-b1e1-b7a2d8808c55.jpg)

I also tried making a custom go2rtc yaml configuration and passed it in docker compose and it didn't make it work remotely. 

```yaml
webrtc:
  listen: ":8555"
  candidates:
    - 192.168.4.40:8555
    - webrtc.mydomain.net

**Comment by spacebares** [comment]
I use cloudflare tunnel with companion app and to get my webrtc working i had to add external router ip to candidates. Did not have to touch any port forwarding. I assume webrtc.mydomain.net is external ip to the network running HA/Frigate?

**Reply by Daniel-dev22** [reply]
It's the CNAME not the IP address. 

Interesting there's 15 possible cloudflare IP addresses, did you use all of them? 

https://www.cloudflare.com/ips-v4

Can you share your configuration?

**Reply by spacebares** [reply]
a domain using name server ```KRISTIN.NS.CLOUDFLARE.COM``` & ```MILES.NS.CLOUDFLARE.COM```
same domain is set as url for the companion app, i dont know if it matters but i have this in HA config:
```
homeassistant:
  external_url: "https://URL.com"
  internal_url: "http://internalIP:8123"
```

a config file called _frigate-go2rtc.yml_ 
```
webrtc:
  listen: ":8555"
  candidates:
    - ExternalIPfromISPmodem:8555 
    - stun:8555
```

config is fed to frigate threw docker-compose as a line in _volumes:_
```      - ./config/frigate-go2rtc.yaml:/config/frigate-go2rtc.yaml:ro```

frigate docker container is set to HOST mode **(important)**, same place docker-compose:
```
services:
  frigate:
    network_mode: host
```


nothing special for the cloudflare tunnel that is running on the machine

**Reply by NickM-27** [reply]
If frigate is set as host and your router supports it, the udp ports will be used with automatic pass through. You can enable webRTC logs in go2rtc config to see which ports are used. 

https://www.homenethowto.com/ports-and-nat/upnp-automatic-port-forward/

It's also worth noting that `ExternalIPfromISPmodem:8555` and `stun:8555` are the same thing. A stun server is just a server to get your hosts public facing ip address. 

https://github.com/AlexxIT/go2rtc#module-webrtc

**Reply by Daniel-dev22** [reply]
> a domain using name server `KRISTIN.NS.CLOUDFLARE.COM` & `MILES.NS.CLOUDFLARE.COM` same domain is set as url for the companion app, i dont know if it matters but i have this in HA config:
> 
> ```
> homeassistant:
>   external_url: "https://URL.com"
>   internal_url: "http://internalIP:8123"
> ```
> 
> a config file called _frigate-go2rtc.yml_
> 
> ```
> webrtc:
>   listen: ":8555"
>   candidates:
>     - ExternalIPfromISPmodem:8555 
>     - stun:8555
> ```
> 
> config is fed to frigate threw docker-compose as a line in _volumes:_ ` - ./config/frigate-go2rtc.yaml:/config/frigate-go2rtc.yaml:ro`
> 
> frigate docker container is set to HOST mode **(important)**, same place docker-compose:
> 
> ```
> services:
>   frigate:
>     network_mode: host
> ```
> 
> nothing special for the cloudflare tunnel that is running on the machine

What firewall are you using? Are you using UPNP?

**Reply by sn0kerbuzz** [reply]
So, is it possible to use the Cloudflare tunnel and webRTC without forwarding any ports? If so, can anyone help me 😄?

**Reply by NickM-27** [reply]
> So, is it possible to use the Cloudflare tunnel and webRTC without forwarding any ports? If so, can anyone help me 😄?

Yes, as long as you don't have a symmetrical NAT

**Reply by sn0kerbuzz** [reply]
I'm a new to this and I don't quite understand what a symmetrical NAT is, but I used this https://tomchen.github.io/symmetric-nat-test/ and it says that I'm have a normal NAT. Will you be willing to walk me through?

**Reply by NickM-27** [reply]
I'd be happy to answer questions and provide details about how I have implemented things, but there are plenty of guides for setting up cloudflare

**Reply by sn0kerbuzz** [reply]
I've set up Cloudflare and I can access Home Assistant on mobile data using the companion app, but I can't manage to make webRTC work.

**Reply by NickM-27** [reply]
If you are not going to be using port forwarding then you need to run frigate in host networking mode

**Reply by sn0kerbuzz** [reply]
I'm using HassOS with the Frigate addon, so instead I will have to install it via Docker, enabling host mode, right?

**Reply by NickM-27** [reply]
Yes, for the addon you'll need to do port forwarding. That being said, I'm not sure what the issue is with port forwarding since it's just port 8555 which will only be communicating with encrypted video / audio data.

**Reply by sn0kerbuzz** [reply]
I didn't know that. So I forwarded port 8555 in my routers and added my external IP in the `candidates` list in `go2rtc.yaml`, but it still doesn't work. Am I still missing something 😢?

**Reply by NickM-27** [reply]
Go2rtc.yaml has nothing to do with frigate, frigates go2rtc config is entirely nested inside the frigate config

**Reply by sn0kerbuzz** [reply]
Yes, in 0.12.0 beta it is nested. I've updated the addon to use v0.12.0 beta 8 and configured the `go2rtc` in the frigate.yml, but same story, webrtc is accessible only on my local network. 😞

**Reply by NickM-27** [reply]
You need to use the logs to figure out why, which part is failing. Also, when configuring go2rtc may be better to use stun:8555 instead of the direct public IP unless your ISP has given you a static IP

**Reply by sn0kerbuzz** [reply]
Thanks, I will look at the logs, see if can find anything.

**Comment by AlexxIT** [comment]
8555 is not HTTP port (OCI level 7). It is TCP port (OCI level 4).

**Reply by Daniel-dev22** [reply]
> 8555 is not HTTP port (OCI level 7). It is TCP port (OCI level 4).

I made that change. I have progress. Semi progress though. I'm getting immediate webrtc errors (progress because it at least says it's trying before I never got these errors)
![Screenshot_20221216_100111_Home Assistant](https://user-images.githubusercontent.com/47092714/208136984-7fcd87d1-7edc-4428-b96e-d64ab9c26b90.jpg)


I'm using the rtsp2webrtc integration pointed at the server running frigate on port 1984. 
`http://192.168.4.40:1984`

I setup the tunnel as you said with tcp.
![Screenshot_20221216_105019_Chrome](https://user-images.githubusercontent.com/47092714/208136951-074942fc-72be-4279-996f-5ce80df450a9.jpg)

I also added in the frigate-go2rtc.yaml

```yaml
webrtc:
  listen: ":8555"
  candidates:
    - 192.168.4.40:8555
    - frigate.mydomain.net
```
I have home assistant set to capture everything I think

```
logger:
 logs:
```

Here's the home assistant logs 
`https://hadomain.net`
This is the domain the companion app uses to connect.


[_homeassistant_logs (6).txt](https://github.com/blakeblackshear/frigate/files/10247215/_homeassistant_logs.6.txt)

**Reply by AlexxIT** [reply]
You don't need local candidate in the list `192.168.4.40:8555`. Because local connection should use random UDP port.

But you need port in your public candidate `frigate.mydomain.net:8555`.

Also default Hass card is not very smart with external WebRTC connection. You can check if go2rtc Web UI version can connect without problems.

**Reply by Daniel-dev22** [reply]
> You don't need local candidate in the list `192.168.4.40:8555`. Because local connection should use random UDP port.
> 
> But you need port in your public candidate `frigate.mydomain.net:8555`.
> 
> Also default Hass card is not very smart with external WebRTC connection. You can check if go2rtc Web UI version can connect without problems.

I made those changes and I'm not getting the webrtc error anymore and webrtc isn't loading but I do see the same errors in the home assistant logs. 

> You can check if go2rtc Web UI version can connect without problems.

As in check if I connect using the tunnel on 1984?

Earlier I also had this in the frigate logs when I was getting those webrtc errors.


```
2022-12-16 14:58:15.590738850  2022/12/16 14:58:15 http: panic serving 192.168.4.40:57032: runtime error: slice bounds out of range [:-1]
2022-12-16 14:58:15.590744649  goroutine 21 [running]:
2022-12-16 14:58:15.590747272  net/http.(*conn).serve.func1()
2022-12-16 14:58:15.590751028  	net/http/server.go:1850 +0xbf
2022-12-16 14:58:15.590754378  panic({0xb34e80, 0xc00052b038})
2022-12-16 14:58:15.590756981  	runtime/panic.go:890 +0x262
2022-12-16 14:58:15.590785729  github.com/AlexxIT/go2rtc/pkg/webrtc.LookupIP({0xc00003c870, 0x24})
2022-12-16 14:58:15.590788315  	github.com/AlexxIT/go2rtc/pkg/webrtc/helper.go:54 +0x228
2022-12-16 14:58:15.590793928  github.com/AlexxIT/go2rtc/cmd/webrtc.syncCanditates({0xc0001f7500, 0x13a7})
2022-12-16 14:58:15.590798263  	github.com/AlexxIT/go2rtc/cmd/webrtc/candidates.go:54 +0x2a8
2022-12-16 14:58:15.590803396  github.com/AlexxIT/go2rtc/cmd/webrtc.ExchangeSDP(0xc000230180, {0xc0002f3800, 0x1746}, {0xc000282201, 0x31})
2022-12-16 14:58:15.590829661  	github.com/AlexxIT/go2rtc/cmd/webrtc/webrtc.go:208 +0x371
2022-12-16 14:58:15.590832336  github.com/AlexxIT/go2rtc/cmd/hass.initAPI.func3({0xc85180, 0xc0002a2380}, 0xc000292800)
2022-12-16 14:58:15.590835891  	github.com/AlexxIT/go2rtc/cmd/hass/api.go:82 +0x2d6
2022-12-16 14:58:15.590838689  net/http.HandlerFunc.ServeHTTP(0xc0002a2380?, {0xc85180?, 0xc0002a2380?}, 0xb7c4ce?)
2022-12-16 14:58:15.590852923  	net/http/server.go:2109 +0x2f
2022-12-16 14:58:15.590855683  net/http.(*ServeMux).ServeHTTP(0x0?, {0xc85180, 0xc0002a2380}, 0xc000292800)
2022-12-16 14:58:15.590857881  	net/http/server.go:2487 +0x149
2022-12-16 14:58:15.590860847  net/http.serverHandler.ServeHTTP({0xc83590?}, {0xc85180, 0xc0002a2380}, 0xc000292800)
2022-12-16 14:58:15.590863189  	net/http/server.go:2947 +0x30c
2022-12-16 14:58:15.590865599  net/http.(*conn).serve(0xc000218140, {0xc859b0, 0xc0001960c0})
2022-12-16 14:58:15.590885323  	net/http/server.go:1991 +0x607
2022-12-16 14:58:15.590887803  created by net/http.(*Server).Serve
2022-12-16 14:58:15.590890079  	net/http/server.go:3102 +0x4db
```

**Reply by Daniel-dev22** [reply]
Should I be using the cloudflare tunnel proxy list listening on :8555 rather than the
`frigate.mydomain.net:8555`

https://www.cloudflare.com/ips-v4

**Reply by AlexxIT** [reply]
`webrtc.LookupIP` because you add candidate without port

**Reply by Daniel-dev22** [reply]
Am I supposed to add `webrtc.LookupIP` somewhere?

I have this.

```yaml
webrtc:
  listen: ":8555"
  candidates:
    - frigate.mydomain.net:8555
```

After I added the port. I stopped seeing both the webrtc error on  the home assistant Frigate card and the go2rtc errors. I'm wondering if it isn't even trying to work now.

No errors in home assistant either that were appearing before.

**Comment by amaciuc** [comment]
@Daniel-dev22 

I tried so many times to configure that tcp socket with cloudflare and to follow other documentations, but nothing ...
My last try was [ngrok](https://github.com/AlexxIT/go2rtc#module-ngrok). If I knew form the beginning that this process is so easy, I wouldn't have tried with cloudflare :smile:

Here are some steps that I followed:
- Create account on `ngrok` website
- It will generate an `authtoken`
- Make a config file `ngrok.yaml` like that:
```
version: "2"
authtoken: <YOUR_AUTHTOKEN>
tunnels:
  webrtc:
    addr: 8555 # use the same port as in go2rtc config
    proto: tcp
```
- Add a new service in `go2rtc.yaml` config file like that:
```
ngrok:
  command: ngrok start --all --config ngrok.yaml
```
- Restart `go2rtc` and let's rock

Now I can view the cameras in hass companion app over data mobile.

**Reply by AlexxIT** [reply]
You don't even need the ngrok configuration file to forward a single port

**Reply by amaciuc** [reply]
Yes, you are right. I complicated it a little with an extra config file.

**Reply by luisiam** [reply]
Do I need to do port forwarding on the router if I use ngrok?

**Reply by NickM-27** [reply]
> Do I need to do port forwarding on the router if I use ngrok?

no you don't

**Reply by luisiam** [reply]
> > Do I need to do port forwarding on the router if I use ngrok?
> 
> no you don't

Hmm, webrtc is still not working remotely. I am using cloudflare tunnel to access my frigate instance. Does that matter?

**Reply by NickM-27** [reply]
I'd suggest making your own discussion and providing info on your setup, config, logs, etc.
